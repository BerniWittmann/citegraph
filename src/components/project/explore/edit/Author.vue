<template>
  <div class="project-explore-single-author-edit">
    <v-card class="pa-4">
      <v-card-title class="headline">
        <h2 class="display-5" v-once>{{ author.firstName }} {{ author.lastName }}</h2>
      </v-card-title>
      <v-container>
        <v-form ref="form" @submit.prevent="submit" class="mt-12" v-model="valid">
          <v-row justify="center">
            <v-col lg="6">
              <v-text-field
                :label="$t('project.explore.edit.author.first_name')"
                v-model="firstName"
                :rules="nameRules"
                required
                autofocus/>
            </v-col>
          </v-row>
          <v-row justify="center">
            <v-col lg="6">
              <v-text-field
                :label="$t('project.explore.edit.author.last_name')"
                v-model="lastName"
                :rules="nameRules"
                required
                autofocus/>
            </v-col>
          </v-row>
          <v-row class="my-10" justify="space-between">
            <v-col cols="auto">
              <v-btn rounded large color="warning" @click="back">
                <v-icon>mdi-menu-left</v-icon>
                {{ $t('project.explore.edit.back') }}
              </v-btn>
            </v-col>

            <v-col cols="auto">
              <v-btn
                rounded large
                :disabled="!valid"
                :loading="loading"
                color="success"
                type="submit"
              >{{ $t('project.explore.edit.submit') }}
              </v-btn>
            </v-col>
          </v-row>
        </v-form>
      </v-container>
    </v-card>
  </div>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator'
import Author from '@/models/paperEntities/author'

@Component
export default class AuthorEditPage extends Vue {
  valid: boolean = true
  loading: boolean = false
  firstName: string = ''
  lastName: string = ''

  nameRules: Array<Function> = [
    (v: string | undefined) => !!v || this.$root.$t('project.explore.edit.author.name_required'),
    (v: string | undefined) => (v && v.trim().length > 0) || this.$root.$t('project.explore.edit.author.name_not_empty')
  ]

  urlRules: Array<Function> = []

  get author (): Author {
    return this.$store.getters['paperEntities/activeEntity'] as Author
  }

  beforeMount (): void {
    this.firstName = this.author.firstName
    this.lastName = this.author.lastName
  }

  async submit (): Promise<void> {
    if (!this.valid) return
    this.loading = true
    await this.$store.dispatch('paperEntities/updateEntity', {
      params: {
        projectId: this.$route.params.projectId,
        entityType: this.$route.params.queryByType,
        id: this.$route.params.entityId
      },
      data: {
        firstName: this.firstName,
        lastName: this.lastName
      }
    })
    this.loading = false
    this.$router.back()
  }

  back (): void {
    this.$router.back()
  }
}
</script>
