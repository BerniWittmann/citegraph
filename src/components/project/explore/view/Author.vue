<template>
  <div class="project-explore-single-author">
    <layout>
      <template v-slot:title>
        <v-avatar color="primary" class="mr-5" size="62">
          <b class="white--text">{{ initials }}</b>
        </v-avatar>
        <h2 class="display-5">{{ author.firstName }} {{ author.lastName }}</h2>
      </template>
      <v-container>
        <v-row>
          <v-col :cols="3">
            <v-list>
              <v-subheader class="title">{{ $t('project.explore.view.author.metrics') }}</v-subheader>
              <v-list-item>
                <v-list-item-content>{{ $t('project.explore.view.author.count_records') }}:</v-list-item-content>
                <v-list-item-content class="align-end">{{ author.countRecords }}</v-list-item-content>
              </v-list-item>
            </v-list>
          </v-col>
        </v-row>
      </v-container>
    </layout>
  </div>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator'
import PaperEntityLayout from '@/layouts/PaperEntity.vue'
import Author from '@/models/paperEntities/author'

@Component({
  components: {
    Layout: PaperEntityLayout
  }
})
export default class AuthorViewPage extends Vue {
  get author (): Author {
    return this.$store.getters['paperEntities/activeEntity'] as Author
  }

  get initials (): string {
    return (this.author.firstName.substring(0, 1) + this.author.lastName.substring(0, 1)).toUpperCase()
  }
}
</script>
