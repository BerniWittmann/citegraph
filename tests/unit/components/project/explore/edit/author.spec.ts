import { shallowMount } from '@vue/test-utils'
import { i18n } from '../../../../setupPlugins'

import EmptySlotComponent from '../../../../EmptySlotComponent.vue'
import AuthorEditPage from '@/components/project/explore/edit/Author.vue'
import Author from '@/models/paperEntities/author'

describe('components/project/explore/edit/Author.vue', () => {
  const router = {
    back: jest.fn()
  }
  const dispatch = jest.fn()

  function getWrapper () {
    return shallowMount(AuthorEditPage, {
      i18n,
      mocks: {
        $store: {
          getters: {
            'paperEntities/activeEntity': new Author({
              id: '1',
              firstName: 'Hans',
              lastName: 'Meier',
              countRecords: 12
            })
          },
          dispatch
        },
        $router: router,
        $route: {
          params: {
            projectId: 42,
            queryByType: 'author',
            entityId: '1'
          }
        },
        $t: (key: string) => key
      },
      stubs: {
        layout: EmptySlotComponent
      }
    })
  }

  beforeEach(() => {
    jest.clearAllMocks()
  })

  it('renders the page', () => {
    const wrapper = getWrapper()
    // @ts-ignore
    wrapper.vm.nameRules = []
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('has a back button', () => {
    const wrapper = getWrapper()
    const btn = wrapper.find('v-btn-stub')
    expect(btn.exists()).toBeTruthy()
    expect(btn.text()).toContain('project.explore.edit.back')
    expect(router.back).not.toHaveBeenCalled()
    btn.vm.$emit('click')
    expect(router.back).toHaveBeenCalled()
  })

  it('has a form', () => {
    const wrapper = getWrapper()
    expect(wrapper.contains('v-form-stub')).toBeTruthy()
  })

  describe('it has a text field of the name', () => {
    it('has a text field', () => {
      const wrapper = getWrapper()
      expect(wrapper.contains('v-text-field-stub')).toBeTruthy()
    })
    it('the text needs to be entered', () => {
      const wrapper = getWrapper()
      // @ts-ignore
      const rules = wrapper.vm.nameRules
      expect.assertions(rules.length)
      for (let rule of rules) {
        expect(rule('test')).toBeTruthy()
      }
    })
    it('the text is required', () => {
      const wrapper = getWrapper()
      // @ts-ignore
      const rule = wrapper.vm.nameRules[0]
      expect(rule(undefined)).toEqual('project.explore.edit.author.name_required')
      expect(rule(null)).toEqual('project.explore.edit.author.name_required')
      expect(rule('')).toEqual('project.explore.edit.author.name_required')
    })
    it('the text can not be an empty string', () => {
      const wrapper = getWrapper()
      // @ts-ignore
      const rule = wrapper.vm.nameRules[1]
      expect(rule('    ')).toEqual('project.explore.edit.author.name_not_empty')
      expect(rule('   test   ')).toBeTruthy()
      expect(rule(' ')).toEqual('project.explore.edit.author.name_not_empty')
    })
  })

  describe('it can submit the form', () => {
    it('saves the author on submit', async () => {
      const wrapper = getWrapper()
      // @ts-ignore
      wrapper.vm.firstName = 'Olli'
      // @ts-ignore
      wrapper.vm.lastName = 'Kahn'
      // @ts-ignore
      await wrapper.vm.submit()
      expect(dispatch).toHaveBeenCalledWith('paperEntities/updateEntity', {
        params: {
          projectId: 42,
          entityType: 'author',
          id: '1'
        },
        data: {
          firstName: 'Olli',
          lastName: 'Kahn'
        }
      })
      expect(router.back).toHaveBeenCalled()
    })

    it('does nothing if the form is not valid', () => {
      const wrapper = getWrapper()
      // @ts-ignore
      wrapper.vm.valid = false
      // @ts-ignore
      wrapper.vm.submit()
      expect(dispatch).not.toHaveBeenCalled()
      expect(router.back).not.toHaveBeenCalled()
    })
  })
})
