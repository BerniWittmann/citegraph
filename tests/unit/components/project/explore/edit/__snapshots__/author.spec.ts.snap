// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/project/explore/edit/Author.vue renders the page 1`] = `
<div class="project-explore-single-author-edit">
  <v-card-stub loaderheight="4" tag="div" class="pa-4">
    <v-card-title-stub class="headline">
      <h2 class="display-5">Hans Meier</h2>
    </v-card-title-stub>
    <v-container-stub tag="div">
      <v-form-stub value="true" class="mt-12">
        <v-row-stub tag="div" justify="center">
          <v-col-stub lg="6" tag="div">
            <v-text-field-stub errorcount="1" errormessages="" messages="" rules="" successmessages="" value="Hans" backgroundcolor="" label="project.explore.edit.author.first_name" loaderheight="2" autofocus="true" clearicon="$clear" type="text" required=""></v-text-field-stub>
          </v-col-stub>
        </v-row-stub>
        <v-row-stub tag="div" justify="center">
          <v-col-stub lg="6" tag="div">
            <v-text-field-stub errorcount="1" errormessages="" messages="" rules="" successmessages="" value="Meier" backgroundcolor="" label="project.explore.edit.author.last_name" loaderheight="2" autofocus="true" clearicon="$clear" type="text" required=""></v-text-field-stub>
          </v-col-stub>
        </v-row-stub>
        <v-row-stub tag="div" justify="space-between" class="my-10">
          <v-col-stub cols="auto" tag="div">
            <v-btn-stub color="warning" tag="button" activeclass="" large="true" rounded="true" type="button">
              <v-icon-stub>mdi-menu-left</v-icon-stub>
              project.explore.edit.back
            </v-btn-stub>
          </v-col-stub>
          <v-col-stub cols="auto" tag="div">
            <v-btn-stub color="success" tag="button" activeclass="" large="true" rounded="true" type="submit">project.explore.edit.submit
            </v-btn-stub>
          </v-col-stub>
        </v-row-stub>
      </v-form-stub>
    </v-container-stub>
  </v-card-stub>
</div>
`;
